solutions/CF Cellular Network.mdx

```diff
@@ -2,62 +2,64 @@
 id: cellular-network
 source: CF
 title: Cellular Network
-author: Benjamin Qi
+author: Benjamin Qi, Maggie Liu, Nathan Wang
 ---
 
-## Solution 1 - Using a Set
-
 [Official Editorial](https://codeforces.com/blog/entry/46324?locale=en)
 
-This solution can also be implemented with Binary Search, but it's better to implement it with a set. The code below is provided only if you want additional practice for binary searching a sorted array.
+<LanguageSection>
+
+<CPPSection>
+
+## Solution 1 - Binary Search on a Sorted Array
 
 ```cpp
 #include <bits/stdc++.h>
 using namespace std;
 
 // returns the first index in the array that is >= value, or arr.size() if no such index exists
```

Comment:
> // returns the first index in the array that is >= value, or arr.size() if no such index exists

this seems wrong ... @thecodingwizard I think you added this?

Reply:
actually it seems right but confusing ... seems odd to end a loop when `lo` *exceeds* `hi`?

Reply:
D: it highkey took me 30 minutes to write that binary search function; our implementations are highly inconsistent across solutions / guide / slides (ex. https://usaco.guide/silver/binary-search-ans?lang=cpp#introduction)

We should standardize binary search impl; should we just do this?

```cpp
int lo = 0, hi = n-1, ans = [defaultAns];

while (lo < hi) {
    int mid = (lo+hi)/2;
    if (f(mid)) lo = mid + 1;
    else hi = mid;
}
```

That was how I learned it at least :/

Reply:
My go-to:

```
int l=-1, r=n; // [0, l]: valid answer --- [r, n]: invalid answers
while(l<r)
{
  int m=(l+r)/2;
  if(ok(m)) l=m;
  else r=m;
}
return r;
```

Reply:
what i wrote ^^^ is equivalent to upper_bound in cpp I believe

Reply:
>  ans = [defaultAns];

why do you need an `ans` variable?

```
int l=-1, r=n; // [0, l]: valid answer --- [r, n]: invalid answers
while(l<r)
{
  int m=(l+r)/2;
  if(ok(m)) l=m;
  else r=m;
}
return r;
```

If the return value is supposed to be `-1` then won't this loop infinitely?

Reply:
> https://usaco.guide/silver/binary-search-ans?lang=cpp#introduction

I think we should just remove the first implementation from here

Reply:
> why do you need an ans variable?

Err I think I typically do `if (ok(mid)) l = mid, ans = mid;` or something like that. It's probably not needed though.

> I think we should just remove the first implementation from here

I think the while loop impl is easier to approach compared to the for loop though?

Reply:
> If the return value is supposed to be -1 then won't this loop infinitely?

yes oops.
I meant `while(l+1<r)`